src/mlpack/bindings/cli/end_program.hpp

```diff
@@ -26,79 +26,69 @@ namespace cli {
 inline void EndProgram()
 {
   // Stop the CLI timers.
-  CLI::GetSingleton().timer.StopAllTimers();
+  IO::GetSingleton().timer.StopAllTimers();
 
   // Print any output.
-  const std::map<std::string, util::ParamData>& parameters = CLI::Parameters();
-  std::map<std::string, util::ParamData>::const_iterator it =
-      parameters.begin();
-  while (it != parameters.end())
+  std::map<std::string, util::ParamData>& parameters = IO::Parameters();
+  for (auto it : parameters)
   {
-    const util::ParamData& d = it->second;
+    util::ParamData& d = it.second;
     if (!d.input)
-      CLI::GetSingleton().functionMap[d.tname]["OutputParam"](d, NULL, NULL);
-
-    ++it;
+      IO::GetSingleton().functionMap[d.tname]["OutputParam"](d, NULL, NULL);
   }
 
-  if (CLI::HasParam("verbose"))
+  if (IO::HasParam("verbose"))
   {
     Log::Info << std::endl << "Execution parameters:" << std::endl;
 
     // Print out all the values.
-    it = parameters.begin();
-    while (it != parameters.end())
+    for (auto it : parameters)
     {
       // Now, figure out what type it is, and print it.
       // We can handle strings, ints, bools, doubles.
-      const util::ParamData& data = it->second;
-      std::string boostName;
-      CLI::GetSingleton().functionMap[data.tname]["MapParameterName"](data,
-          NULL, (void*) &boostName);
-      Log::Info << "  " << boostName << ": ";
+      util::ParamData& data = it.second;
+      std::string cliName;
+      IO::GetSingleton().functionMap[data.tname]["MapParameterName"](data,
+          NULL, (void*) &cliName);
+      Log::Info << "  " << cliName << ": ";
 
       std::string printableParam;
-      CLI::GetSingleton().functionMap[data.tname]["GetPrintableParam"](data,
+      IO::GetSingleton().functionMap[data.tname]["GetPrintableParam"](data,
           NULL, (void*) &printableParam);
       Log::Info << printableParam << std::endl;
-
-      ++it;
     }
 
     Log::Info << "Program timers:" << std::endl;
-    for (auto it2 : CLI::GetSingleton().timer.GetAllTimers())
+    for (auto it2 : IO::GetSingleton().timer.GetAllTimers())
     {
       Log::Info << "  " << it2.first << ": ";
-      CLI::GetSingleton().timer.PrintTimer(it2.first);
+      IO::GetSingleton().timer.PrintTimer(it2.first);
     }
   }
 
   // Lastly clean up any memory.  If we are holding any pointers, then we "own"
   // them.  But we may hold the same pointer twice, so we have to be careful to
   // not delete it multiple times.
-  std::unordered_map<void*, const util::ParamData*> memoryAddresses;
-  it = parameters.begin();
-  while (it != parameters.end())
+  std::unordered_map<void*, util::ParamData*> memoryAddresses;
+  for (auto it : parameters)
```

Comment:
```suggestion
  for (auto& it : parameters)
```
This is why the Windows build fails.  When we do `auto it : parameters`, this does not use references---it creates a copy of each element in `parameters`.  Thus when we push `&data` onto the map `memoryAddresses`, that memory address goes out of scope as soon as we move on to the next loop iteration.  Therefore in the next loop, when we iterate over `memoryAddresses`, all the addresses are invalid.  If you want to see, take a look at the build logs and the changes in #2505.  If we instead use `auto& it : parameters`, this uses a reference to each element in `parameters` instead of making a copy, and the code works fine.

I'll go throughout and make suggestions to replace the `auto it` pattern with `auto& it`. :+1: